From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@Gmail.com>
Date: Tue, 18 Jan 2022 06:35:54 -0600
Subject: [PATCH] Purpur: Max joins per second

Original license: MIT
Original project: https://github.com/PurpurMC/Purpur

When this option is set to true the `max-joins-per-tick` setting in paper.yml will be used per second instead of per tick

diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index b6a80edbbceb85932a5fa8b9488d44d41f5c5314..ef04e642856358cf633bcdbb923051224c54f1d6 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -42,6 +42,7 @@ import net.minecraft.server.RunningOnDifferentThreadException;
 import net.minecraft.util.LazyLoadedValue;
 import net.minecraft.util.Mth;
 import org.apache.commons.lang3.Validate;
+import org.dreeam.leaf.LeafConfig;
 import org.slf4j.Logger;
 import org.slf4j.Marker;
 import org.slf4j.MarkerFactory;
@@ -562,11 +563,20 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     private static final int MAX_PER_TICK = io.papermc.paper.configuration.GlobalConfiguration.get().misc.maxJoinsPerTick; // Paper
     private static int joinAttemptsThisTick; // Paper
     private static int currTick; // Paper
+    private static int tickSecond; // Purpur
     public void tick() {
         this.flushQueue();
         // Paper start
         if (Connection.currTick != net.minecraft.server.MinecraftServer.currentTick) {
             Connection.currTick = net.minecraft.server.MinecraftServer.currentTick;
+            // Purpur start
+            if (LeafConfig.maxJoinsPerSecond) {
+                if (++Connection.tickSecond > 20) {
+                    Connection.tickSecond = 0;
+                    Connection.joinAttemptsThisTick = 0;
+                }
+            } else
+            // Purpur end
             Connection.joinAttemptsThisTick = 0;
         }
         // Paper end
diff --git a/src/main/java/org/dreeam/leaf/LeafConfig.java b/src/main/java/org/dreeam/leaf/LeafConfig.java
index 1173f414308446fb0c93076896f0431cc7c92aa6..a6c1143412c6ccfa7e0c83d44b30c72c085fca33 100644
--- a/src/main/java/org/dreeam/leaf/LeafConfig.java
+++ b/src/main/java/org/dreeam/leaf/LeafConfig.java
@@ -466,8 +466,10 @@ public class LeafConfig {
     }
 
     public static boolean kickForOutOfOrderChat = true;
+    public static boolean maxJoinsPerSecond = false;
     private static void networkSettings() {
         kickForOutOfOrderChat = getBoolean("network.kick-for-out-of-order-chat", kickForOutOfOrderChat);
+        maxJoinsPerSecond = getBoolean("network.max-joins-per-second", maxJoinsPerSecond);
     }
 
     public static double axolotlMaxHealth = 14.0D;
